pkg/scheduler/framework/plugins/podtopologyspread/scoring.go

```diff
@@ -0,0 +1,255 @@
+					matchSum++
+				}
+			}
+			atomic.AddInt64(state.topologyPairToPodCounts[pair], matchSum)
+		}
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodes), processAllNode)
+
+	cycleState.Write(postFilterStateKey, state)
+	return nil
+}
+
+// Score invoked at the Score extension point.
+// The "score" returned in this function is the matching number of pods on the `nodeName`,
+// it is normalized later.
+func (pl *PodTopologySpread) Score(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	nodeInfo, err := pl.sharedLister.NodeInfos().Get(nodeName)
+	if err != nil || nodeInfo.Node() == nil {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v, node is nil: %v", nodeName, err, nodeInfo.Node() == nil))
+	}
+
+	node := nodeInfo.Node()
+	s, err := getPostFilterState(cycleState)
+	if err != nil {
+		return 0, framework.NewStatus(framework.Error, err.Error())
+	}
+
+	// Return if the node is not qualified.
+	if _, ok := s.nodeNameSet[node.Name]; !ok {
+		return 0, nil
+	}
+
+	// For each present <pair>, current node gets a credit of <matchSum>.
+	// And we sum up <matchSum> and return it as this node's score.
+	var score int64
+	for _, c := range s.constraints {
+		if tpVal, ok := node.Labels[c.topologyKey]; ok {
+			pair := topologyPair{key: c.topologyKey, value: tpVal}
+			matchSum := *s.topologyPairToPodCounts[pair]
+			score += matchSum
+		}
+	}
+	return score, nil
+}
+
+// NormalizeScore invoked after scoring all nodes.
+func (pl *PodTopologySpread) NormalizeScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, scores framework.NodeScoreList) *framework.Status {
+	s, err := getPostFilterState(cycleState)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	if s == nil {
+		return nil
+	}
+
+	// Calculate the summed <total> score and <minScore>.
+	var minScore int64 = math.MaxInt64
+	var total int64
+	for _, score := range scores {
+		// it's mandatory to check if <score.Name> is present in m.nodeNameSet
+		if _, ok := s.nodeNameSet[score.Name]; !ok {
+			continue
+		}
+		total += score.Score
+		if score.Score < minScore {
+			minScore = score.Score
+		}
+	}
+
+	maxMinDiff := total - minScore
+	for i := range scores {
+		nodeInfo, err := pl.sharedLister.NodeInfos().Get(scores[i].Name)
+		if err != nil {
+			return framework.NewStatus(framework.Error, err.Error())
+		}
+		node := nodeInfo.Node()
+		// Debugging purpose: print the score for each node.
+		// Score must be a pointer here, otherwise it's always 0.
+		if klog.V(10) {
+			defer func(score *int64, nodeName string) {
+				klog.Infof("%v -> %v: PodTopologySpread NormalizeScore, Score: (%d)", pod.Name, nodeName, *score)
+			}(&scores[i].Score, node.Name)
+		}
+
+		if maxMinDiff == 0 {
+			scores[i].Score = framework.MaxNodeScore
+			continue
+		}
+
+		if _, ok := s.nodeNameSet[node.Name]; !ok {
+			scores[i].Score = 0
```

Comment:
shouldn't this node have a score of zero already? why do we need to set it explicitly?

Reply:
Technically this line can be removed, however, explicitly setting it to 0 doesn't hurt and make the code easy to read.